tests/main/session-tool/task.yaml

```
@@ -1,23 +1,41 @@
 summary: tests for session-tool
 # Session-tool depends on busctl and doesn't work on 14.04
 systems: [-ubuntu-14.04-*]
-kill-timeout: 1m
+environment:
+    USER/root: root
+    USER/test: test
+prepare: |
+    # Check which sessions we have before running the test
+    loginctl list-sessions > before.txt
+    # Prepare for using sessions as the given user
+    session-tool --prepare -u "$USER"
 execute: |
-    # User switching really works
-    session-tool -u test id -u | MATCH 12345
-    # We get XDG_RUNTIME_DIR
-    session-tool -u test env | MATCH XDG_RUNTIME_DIR=/run/user/12345
-    # We get a logind session
-    session-tool -u test loginctl show-user test | MATCH State=active
-    # Exit code is forwarded
-    session-tool -u test true
-    not session-tool -u test false
+    for n in $(seq 300); do
+        echo "ITERATION $(date) $n"
+        session-tool -u "$USER" id -u  2>/tmp/session-tool.log | MATCH "$(id -u "$USER")"
+        session-tool -u "$USER" env    2>/tmp/session-tool.log | MATCH "XDG_RUNTIME_DIR=/run/user/$(id -u "$USER")"
+        # We get a logind session
+        session-tool -u "$USER" loginctl list-sessions | grep "$USER"
+        # Exit code is forwarded
+        session-tool -u "$USER" true
+        not session-tool -u "$USER" false
+        # The -p option can be used to know the PID of the started program.
+        # This is different from the pid of session-tool as there's a session
+        # manager overlooking the termination of PAM stack (internally we use runuser).
+        session-tool -u "$USER" -p /tmp/outer.pid "$(command -v any-python)" -c 'from __future__ import print_function; import os; print(os.getpid())' >/tmp/inner.pid
+        test "$(cat /tmp/outer.pid)" = "$(cat /tmp/inner.pid)"
+    done
 restore: |
-    # Session termination runs asynchronously. Session services are
-    # automatically terminated when the session is shutting down but this is
-    # not synchronized with the execution of the main test program.
-    #
-    # Notably kills: dbus-daemon, systemd --user.
-    # On desktop systems this might also kill: pulseaudio, bluetoothd, gpg
-    # daemon, apt daemon and more.
-    loginctl kill-user test || :
+    # Restore after using sessions as the given user
+    session-tool --restore -u "$USER"
+
+    # NOTE: This part of the test is very flaky if you restart systemd-logind
+    # It *seems* that all new systems (Debian Sid, Arch, Tumbleweed and Ubuntu 20.04)
+    # Forget about the root session that was there to begin with.
```

I meant to continue the previous sentence here.
```suggestion
    # forget about the root session that was there to begin with.
```